collector/connector/swotracemetricsconnector/connector.go

```diff
@@ -0,0 +1,118 @@
+package swotracemetricsconnector
+
+import (
+	"context"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+	"go.uber.org/zap"
+	"time"
+)
+
+type connectorImp struct {
+	config          Config
+	metricsConsumer consumer.Metrics
+	logger          *zap.Logger
+	component.StartFunc
+	component.ShutdownFunc
+}
+
+func newConnector(logger *zap.Logger, config component.Config) (*connectorImp, error) {
+	logger.Info("Building swo-trace-metrics-connector")
+	cfg := config.(*Config)
+	return &connectorImp{
+		config: *cfg,
+		logger: logger,
+	}, nil
+}
+
+func (con *connectorImp) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{MutatesData: false}
+}
+
+func (con *connectorImp) ConsumeTraces(ctx context.Context, traceData ptrace.Traces) error {
+	for i := 0; i < traceData.ResourceSpans().Len(); i++ {
+		resourceSpan := traceData.ResourceSpans().At(i)
+
+		for j := 0; j < resourceSpan.ScopeSpans().Len(); j++ {
+			scopeSpan := resourceSpan.ScopeSpans().At(j)
+
+			for k := 0; k < scopeSpan.Spans().Len(); k++ {
+				span := scopeSpan.Spans().At(k)
+
+				if span.ParentSpanID().IsEmpty() {
+					startTime := span.StartTimestamp()
+					endTime := span.EndTimestamp()
+
+					if endTime > startTime {
+						duration := uint64(endTime-startTime) / 1000
+						metrics := createMetric(&resourceSpan, &span, duration)
+						if err := con.metricsConsumer.ConsumeMetrics(ctx, metrics); err != nil {
+							con.logger.Error("Failed ConsumeMetrics", zap.Error(err))
+						}
+						return nil
```

Comment:
It seems like returning early might miss some root span since there can be multiple roots in the payload?

Reply:
yes, it is possible.